### ISSUE TITLE
Segfault with OCaml 5

### ISSUE THREAD

**Issue by bensmrs** [astrada/ocamlfuse#26]
Hi!
I don’t know if it’s related to #24, but ocamlfuse (2.7.1_cvs9) crashes with a SIGSEGV with OCaml 5 on my install (Debian Bullseye). With your `hello.ml` example, here is the output from strace:
```
root@comtest-1:/home/bsomers/fuse-test# strace -f _build/default/src/main.exe /tmp/foo
execve("_build/default/src/main.exe", ["_build/default/src/main.exe", "/tmp/foo"], 0x7ffe75cf8240 /* 25 vars */) = 0
brk(NULL)                               = 0x55e2c0740000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (Aucun fichier ou dossier de ce type)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=37803, ...}) = 0
mmap(NULL, 37803, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f1b9508a000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libfuse.so.2", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0p\217\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=260216, ...}) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f1b95088000
mmap(NULL, 262432, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f1b95047000
mprotect(0x7f1b9504f000, 155648, PROT_NONE) = 0
mmap(0x7f1b9504f000, 110592, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x8000) = 0x7f1b9504f000
mmap(0x7f1b9506a000, 40960, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x23000) = 0x7f1b9506a000
mmap(0x7f1b95075000, 77824, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x2d000) = 0x7f1b95075000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libm.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0\362\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=1321344, ...}) = 0
mmap(NULL, 1323280, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f1b94f03000
mmap(0x7f1b94f12000, 630784, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0xf000) = 0x7f1b94f12000
mmap(0x7f1b94fac000, 626688, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0xa9000) = 0x7f1b94fac000
mmap(0x7f1b95045000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x141000) = 0x7f1b95045000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libdl.so.2", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0000\21\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=18688, ...}) = 0
mmap(NULL, 20752, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f1b94efd000
mmap(0x7f1b94efe000, 8192, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1000) = 0x7f1b94efe000
mmap(0x7f1b94f00000, 4096, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3000) = 0x7f1b94f00000
mmap(0x7f1b94f01000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x3000) = 0x7f1b94f01000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libpthread.so.0", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0 |\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=149520, ...}) = 0
mmap(NULL, 136304, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f1b94edb000
mmap(0x7f1b94ee2000, 65536, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x7000) = 0x7f1b94ee2000
mmap(0x7f1b94ef2000, 20480, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x17000) = 0x7f1b94ef2000
mmap(0x7f1b94ef7000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1b000) = 0x7f1b94ef7000
mmap(0x7f1b94ef9000, 13424, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f1b94ef9000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0@n\2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=1839792, ...}) = 0
mmap(NULL, 1852680, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f1b94d16000
mprotect(0x7f1b94d3b000, 1662976, PROT_NONE) = 0
mmap(0x7f1b94d3b000, 1355776, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x25000) = 0x7f1b94d3b000
mmap(0x7f1b94e86000, 303104, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x170000) = 0x7f1b94e86000
mmap(0x7f1b94ed1000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1ba000) = 0x7f1b94ed1000
mmap(0x7f1b94ed7000, 13576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f1b94ed7000
close(3)                                = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f1b94d14000
arch_prctl(ARCH_SET_FS, 0x7f1b94d15440) = 0
mprotect(0x7f1b94ed1000, 12288, PROT_READ) = 0
mprotect(0x7f1b94ef7000, 4096, PROT_READ) = 0
mprotect(0x7f1b94f01000, 4096, PROT_READ) = 0
mprotect(0x7f1b95045000, 4096, PROT_READ) = 0
mprotect(0x7f1b95075000, 73728, PROT_READ) = 0
mprotect(0x55e2bf6f1000, 4096, PROT_READ) = 0
mprotect(0x7f1b950be000, 4096, PROT_READ) = 0
munmap(0x7f1b9508a000, 37803)           = 0
set_tid_address(0x7f1b94d15710)         = 2180479
set_robust_list(0x7f1b94d15720, 24)     = 0
rt_sigaction(SIGRTMIN, {sa_handler=0x7f1b94ee2690, sa_mask=[], sa_flags=SA_RESTORER|SA_SIGINFO, sa_restorer=0x7f1b94eef140}, NULL, 8) = 0
rt_sigaction(SIGRT_1, {sa_handler=0x7f1b94ee2730, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_SIGINFO, sa_restorer=0x7f1b94eef140}, NULL, 8) = 0
rt_sigprocmask(SIG_UNBLOCK, [RTMIN RT_1], NULL, 8) = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
brk(NULL)                               = 0x55e2c0740000
brk(0x55e2c0761000)                     = 0x55e2c0761000
access("/tmp/foo", F_OK)                = 0
lstat("/tmp", {st_mode=S_IFDIR|S_ISVTX|0777, st_size=4096, ...}) = 0
lstat("/tmp/foo", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
pipe([3, 4])                            = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f1b94d15710) = 2180480
read(3, strace: Process 2180480 attached
 <unfinished ...>
[pid 2180480] set_robust_list(0x7f1b94d15720, 24) = 0
[pid 2180480] setsid()                  = 2180480
[pid 2180480] chdir("/")                = 0
[pid 2180480] openat(AT_FDCWD, "/dev/null", O_RDWR) = 5
[pid 2180480] dup2(5, 0)                = 0
[pid 2180480] dup2(5, 1)                = 1
[pid 2180480] dup2(5, 2)                = 2
[pid 2180480] close(5)                  = 0
[pid 2180480] write(4, "\1", 1 <unfinished ...>
[pid 2180479] <... read resumed>"\1", 1) = 1
[pid 2180480] <... write resumed>)      = 1
[pid 2180479] exit_group(0 <unfinished ...>
[pid 2180480] close(3 <unfinished ...>
[pid 2180479] <... exit_group resumed>) = ?
[pid 2180480] <... close resumed>)      = 0
[pid 2180480] close(4)                  = 0
[pid 2180479] +++ exited with 0 +++
mmap(NULL, 268439552, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f1b84d13000
munmap(0x7f1b94d13000, 4096)            = 0
brk(0x55e2c0782000)                     = 0x55e2c0782000
mmap(0x7f1b84d13000, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f1b84d13000
mmap(NULL, 557056, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f1b84c8b000
munmap(0x7f1b84c8b000, 20480)           = 0
munmap(0x7f1b84d10000, 12288)           = 0
sigaltstack({ss_sp=0x55e2c07799b0, ss_flags=0, ss_size=8192}, NULL) = 0
rt_sigaction(SIGPROF, NULL, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
readlink("/proc/self/exe", "/home/bsomers/fuse-test/_build/d"..., 256) = 51
stat("/home/bsomers/fuse-test/_build/default/src/main.exe", {st_mode=S_IFREG|0555, st_size=3704208, ...}) = 0
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x657860} ---
+++ killed by SIGSEGV +++
```
I don’t really know what else I can do to further isolate the problem. The program runs fine in the foreground; the problem only appears when run in the background.

**Comment by astrada** [comment]
Could you please open the core dump with gdb and paste the backtrace here? Thanks! Because from the strace, I can't see the problem (#24 caused a SIGABRT).

**Comment by bensmrs** [comment]
I did some digging and… wow that’s confusing. The program seems to crash depending on its location on the disk. I think I found a pattern, but that sounds a little ridiculous. Could you compile your `hello.ml` example, and move it to the file `/tmp/xxxxx_xxx/_xxxx/xxxx`, then run it to mount the filesystem somewhere?
Here is what I did on my machine: https://asciinema.org/a/y87pVpohYZlQmyNaZMxcKxgPu
Can you reproduce using these exact commands?

**Comment by astrada** [comment]
I was able to reproduce the error. I forgot to set the last element of the copy of `argv` to `NULL` (I'm not very good at C :sweat: ). Should be fixed in version `2.7.1_cvs10`.
